File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/oauth2/docs/github/GithubAuthenticationMapper.kt
package io.micronaut.security.oauth2.docs.github

//tag::clazz[]
import io.micronaut.context.annotation.Requires
import io.micronaut.security.authentication.AuthenticationResponse
import io.micronaut.security.oauth2.endpoint.authorization.state.State
import io.micronaut.security.oauth2.endpoint.token.response.OauthAuthenticationMapper
import io.micronaut.security.oauth2.endpoint.token.response.TokenResponse
import jakarta.inject.Named
import jakarta.inject.Singleton
import org.reactivestreams.Publisher
import reactor.core.publisher.Flux

@Named("github") // <1>
//end::clazz[]
@Requires(property = "docs.classes")
//tag::clazz[]
@Singleton
internal class GithubAuthenticationMapper(private val apiClient: GithubApiClient) // <2>
    : OauthAuthenticationMapper {

    override fun createAuthenticationResponse(tokenResponse: TokenResponse, state: State?): Publisher<AuthenticationResponse> { // <3>
        return Flux.from(apiClient.getUser("token " + tokenResponse.accessToken))
                .map { user ->
                    AuthenticationResponse.success(user.login, listOf("ROLE_GITHUB")) // <4>
                }
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/security/oauth2/docs/github/GithubAuthenticationMapper.py
# tag::clazz[]
from jakarta.inject import Named, Singleton
from micronaut.context.annotation import Requires
from micronaut.security.authentication import AuthenticationResponse
from micronaut.security.oauth2.endpoint.authorization.state import State
from micronaut.security.oauth2.endpoint.token.response import OauthAuthenticationMapper, TokenResponse
from org.reactivestreams import Publisher
from reactor.core.publisher import Flux

from .GithubApiClient import GithubApiClient


@Named("github")  # <1>
# end::clazz[]
@Requires(property="docs.classes")
# tag::clazz[]
@Singleton
class GithubAuthenticationMapper(OauthAuthenticationMapper):

    def __init__(self, apiClient: GithubApiClient):  # <2>
        self.apiClient = apiClient

    def createAuthenticationResponse(self, tokenResponse: TokenResponse, state: State | None) -> Publisher[AuthenticationResponse]:  # <3>
        return Flux.from_(self.apiClient.get_user("token " + tokenResponse.getAccessToken())).map(
            lambda user: AuthenticationResponse.success(user.login, ["ROLE_GITHUB"])  # <4>
        )
# end::clazz[]
